### PR TITLE
chore: tensorflow 2.12 [MLG-358]

### PR DESCRIPTION
**Pull request by drh-determined-ai**
## Description

These changes make Determined compatible with Tensorflow 2.12.0.

Other dependency updates:
- horovod==0.27.0
- jax==0.3.15
- protobuf is no longer pinned for MacOS
- python 3.8.16 (in CircleCI - see note below)

Tensorflow 2.12 requires Python 3.8. In CircleCI, the default version of Python was changed to reflect this. If this is too wide a change, we can revert this and specify 3.8 in workflows that need it (because of TF 2.12).

We are still supporting/using older versions of Tensorflow, and no migrations have been made (yet). In other words, we are using legacy Tensorflow/Keras code, particularly `optimizers.legacy` and `saving.legacy.hdf5_format`. This is necessary due to changes in Tensorflow 2.11.0.

The example `gbt_titanic_estimator` has been disabled for now. The model within uses `tf.estimator.BoostedTreesClassifier` which was removed in Tensorflow 2.9.0. The replacement is the Decision Forest, and the example needs to be updated.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: e2e_tests/tests/requirements.txt
Comment (wes-turner) [inline e2e_tests/tests/requirements.txt:1]: I think the `#tensorflow-metal...` line is accidentally leftover.

File: e2e_tests/tests/requirements.txt [hunk 1]
```diff
@@ -1,14 +1,15 @@
 appdirs
 # TODO(MLG-336): remove this requirement when tensorflow-macos is upgraded.
```
Comment (wes-turner) [inline e2e_tests/tests/requirements.txt:2]: I assume this TODO can be resolved.

The new version restrictions might also be worth a short comment with reasoning.

File: helm/charts/determined/templates/master-config.yaml [hunk 2]
```diff
@@ -131,8 +131,8 @@ data:
       {{- toYaml .Values.resourcePools | nindent 6}}
     {{- end }}
 
-    {{$cpuImage := (split "/" "determinedai/environments:py-3.8-pytorch-1.12-tf-2.8-cpu-6218891")._1}}
```
Comment (wes-turner) [inline helm/charts/determined/templates/master-config.yaml:134]: Should `change-password.yaml` also be updated?

File: .circleci/config.yml [hunk 3]
```diff
@@ -304,7 +304,7 @@ commands:
         type: string
       python-version:
         type: string
-        default: "3.7.15"
+        default: "3.8.16"
```
Comment (wes-turner) [inline .circleci/config.yml:307]: non-blocking: I can't think of any risks here from bumping python. ...can you? This _could_ be done in a separate commit if you're at all worried of risk.

File: e2e_tests/tests/config.py [hunk 4]
```diff
@@ -14,9 +14,9 @@
 
 
 DEFAULT_TF1_CPU_IMAGE = "determinedai/environments:py-3.7-pytorch-1.7-tf-1.15-cpu-6218891"
-DEFAULT_TF2_CPU_IMAGE = "determinedai/environments:py-3.8-pytorch-1.12-tf-2.8-cpu-6218891"
+DEFAULT_TF2_CPU_IMAGE = "determinedai/environments-dev:py-3.8-pytorch-1.12-tf-2.12-cpu-0.21.2"
```
Comment (wes-turner) [inline e2e_tests/tests/config.py:17]: Why is the temporary use of `determinedai/environments-dev` useful? Is there a concern about publishing a 2.12-tagged image to `determinedai/environments`?

File: e2e_tests/tests/config.py [hunk 5]
```diff
@@ -14,9 +14,9 @@
...
 DEFAULT_TF1_GPU_IMAGE = "determinedai/environments:cuda-10.2-pytorch-1.7-tf-1.15-gpu-6218891"
-DEFAULT_TF2_GPU_IMAGE = "determinedai/environments:cuda-11.3-pytorch-1.12-tf-2.8-gpu-6218891"
```
Comment (wes-turner) [inline e2e_tests/tests/config.py:19]: non-blocking: what does the `6218891` represent? Is it a naming convention we've moved away from?

File: requirements.txt [hunk 6]
```diff
@@ -21,6 +21,7 @@ flake8-docstrings>=1.4.0
 flake8-quotes>=2.1.0
 flake8-tuple>=0.4.0
 isort==5.9.3
+jax==0.3.15  # Needed by TF 2.12. Versions 0.4.* use self-documenting f-strings, which are incompatible with Python 3.7. Later patches of 0.3.* cause other mypy errors.
```
Comment (wes-turner) [inline requirements.txt:24]: Is this a good time to unpin click?

File: examples/tests/requirements.txt [hunk 7]
```diff
@@ -1,9 +1,10 @@
 # pytest 6.0 has linter-breaking changes
 # TODO(MLG-336): remove this requirement when tensorflow-macos is upgraded.
-protobuf<3.20; sys_platform == 'darwin' and platform_machine == 'arm64'
+protobuf>=3.20.3,<4.21; sys_platform == 'darwin' and platform_machine == 'arm64'
```
Comment (ioga) [inline examples/tests/requirements.txt:3]: this ends up being equivalent to `protobuf==3.20.3`, no?

File: examples/computer_vision/cifar10_tf_keras/cifar_model.py [hunk 8]
```diff
@@ -6,11 +6,16 @@
...
 import tensorflow as tf
 from data import NUM_CLASSES
+from packaging import version
 from tensorflow.keras.layers import Activation, Conv2D, Dense, Dropout, Flatten, MaxPooling2D
 from tensorflow.keras.losses import categorical_crossentropy
 from tensorflow.keras.metrics import categorical_accuracy
 from tensorflow.keras.models import Sequential
-from tensorflow.keras.optimizers import RMSprop
+
+if version.parse("2.11.0") <= version.parse(tf.__version__):
```
Comment (wes-turner) [inline examples/computer_vision/cifar10_tf_keras/cifar_model.py:15]: nit: It's confusing to me to read a conditional as:
```
if 10 < x:
```
rather than 
```
if x > 10:
```
But I can't find anything in the style guide that disallows it, so to each their own. If this was intentional or an idiom I'm unfamiliar with, then please ignore my comment. If an artifact of the dev process and unintentional, here's a nudge to change it.

File: examples/computer_vision/cifar10_tf_keras/cifar_model.py [hunk 9]
```diff
@@ -6,11 +6,16 @@
...
+    from tensorflow.keras.optimizers.legacy import RMSprop
+else:
+    from tensorflow.keras.optimizers import RMSprop
```
Comment (wes-turner) [inline examples/computer_vision/cifar10_tf_keras/cifar_model.py:18]: I like your suggestion to create a new ticket to migrate to the modern Keras interface. If you do, this is a good spot for a TODO(MLG-XXX):

File: requirements.txt (hunk 6 above)
Comment (wes-turner) [reply requirements.txt:24]: non-blocking: Could this also be `jax<=0.3.15`?

File: schemas/test_cases/v0/experiment.yaml [hunk 10]
```diff
@@ -32,9 +32,9 @@
       environment_variables: {}
       force_pull_image: false
       image:
-        cpu: determinedai/environments:py-3.8-pytorch-1.12-tf-2.8-cpu-6218891
-        cuda: determinedai/environments:cuda-11.3-pytorch-1.12-tf-2.8-gpu-6218891
-        rocm: determinedai/environments:rocm-5.0-pytorch-1.10-tf-2.7-rocm-6218891
+        cpu: determinedai/environments-dev:py-3.8-pytorch-1.12-tf-2.12-cpu-0.21.2
+        cuda: determinedai/environments-dev:cuda-11.3-pytorch-1.12-tf-2.12-gpu-0.21.2
+        rocm: determinedai/environments:rocm-5.0-pytorch-1.10-tf-2.7-rocm-9d07809
```
Comment (wes-turner) [inline schemas/test_cases/v0/experiment.yaml:37]: question: Just for my own chance to learn, could you tell me what this change to `rocm` is for?

**Review by wes-turner — COMMENTED** [review]
nit: Could you please add a reference to [MLG-442] in the commit message explaining the change to `gbt_titanic_estimator`?

If it fits with common practice, could you please also include a sentence or two in the Test Plan section to release partiers about what they should/shouldn't worry about? (if not and we don't leave those kinds of notes, don't worry about it)

Broad strokes, I like the patch and your plan for rollout. Thanks for chasing down and solving all the various TF version locations and incompatibilities. This will be nice to have.

File: examples/tests/requirements.txt (hunk 7 above)
Comment (drh-determined-ai) [reply examples/tests/requirements.txt:3]: I knew you'd ask about this line. :-) I believe you're correct. I'm open to being totally strict here.

File: e2e_tests/tests/config.py (hunk 4 above)
Comment (drh-determined-ai) [reply e2e_tests/tests/config.py:17]: It's just because my PR that's creating these images is still open: https://github.com/determined-ai/environments/pull/196

File: e2e_tests/tests/config.py (hunk 5 above)
Comment (drh-determined-ai) [reply e2e_tests/tests/config.py:19]: That's a commit hash in the `environments` repo. It's a finer granularity than using the Determined version.

File: requirements.txt (hunk 6 above)
Comment (drh-determined-ai) [reply requirements.txt:24]: Nope. Tensorflow 2.12 requires `jax>=0.3.15`. I pinned it at `0.3.15` because later versions caused weird errors with mypy.

File: schemas/test_cases/v0/experiment.yaml (hunk 10 above)
Comment (drh-determined-ai) [reply schemas/test_cases/v0/experiment.yaml:37]: Oops, that is a mistake! Dunno how that happened - overzealous pasting, I suppose.

Comment (drh-determined-ai) [reply schemas/test_cases/v0/experiment.yaml:37]: This change will be reverted before I merge, though, so it's not a worry.

File: .circleci/config.yml (hunk 3 above)
Comment (drh-determined-ai) [reply .circleci/config.yml:307]: The only thing I know of that's holding us to 3.7 is that we're running `mypy` assuming Python 3.7 (see the `mypy.ini` files).

File: helm/charts/determined/templates/master-config.yaml (hunk 2 above)
Comment (drh-determined-ai) [reply helm/charts/determined/templates/master-config.yaml:134]: Oops, I missed that one (I think I only looked in `harness`). I don't know what that configuration is for, beyond the obvious (to change a password :-p), so I dunno how Tensorflow has anything to do with it.
All of these changes to image names are going to be reverted before I do the merge anyway (at least some will be properly updated with a bumpenv later).

File: e2e_tests/tests/requirements.txt
Comment (drh-determined-ai) [reply e2e_tests/tests/requirements.txt:1]: Correct. That switch will be handled in a later issue & PR.

File: e2e_tests/tests/requirements.txt (hunk 1 above)
Comment (wes-turner) [reply e2e_tests/tests/requirements.txt]: Could you please add a short comment with the reasoning behind the pin for us to know later on if and when it should be touched? It's at the moment not clear to me if it wants the "darwin, etc." conditional.

File: e2e_tests/tests/config.py (hunk 4 above)
Comment (wes-turner) [reply e2e_tests/tests/config.py:17]: Will you commit the other before committing this one? If so, then could we amend this PR to have the final path?

File: e2e_tests/tests/config.py (hunk 5 above)
Comment (wes-turner) [reply e2e_tests/tests/config.py:19]: Thanks!

File: requirements.txt (hunk 6 above)
Comment (wes-turner) [reply requirements.txt:24]: Thanks. I hadn't understood from the existing comment.

File: schemas/test_cases/v0/experiment.yaml (hunk 10 above)
Comment (wes-turner) [reply schemas/test_cases/v0/experiment.yaml:37]: Leaving this open because there's something I think I don't understand about the release plan, but that may become obvious shortly.

File: requirements.txt (hunk 6 above)
Comment (wes-turner) [reply requirements.txt:24]: Could you please take a look at the "click" pin and decide what you want to do with it?

File: e2e_tests/tests/config.py (hunk 4 above)
Comment (drh-determined-ai) [reply e2e_tests/tests/config.py]: This PR and the environments PR can be merged in any order. Image names such as these will be changed in a bumpenv PR after both of the aforementioned are merged. Any necessary changes to the CircleCI workflows will also be done in that bumpenv PR.

File: .circleci/config.yml (hunk 3 above)
Comment (drh-determined-ai) [reply .circleci/config.yml]: I _think_ that `python_version` just tells `mypy` what set of exact rules to use when checking code (like whether `dict` is ok in type-hinting).

File: e2e_tests/tests/requirements.txt (hunk 1 above)
Comment (drh-determined-ai) [reply e2e_tests/tests/requirements.txt]: The conditional is important for distinguishing MacOS machines. I don't think we want to mess with that. Switching from `tensorflow-macos` to `tensorflow-metal` will be done later and separately.

File: schemas/test_cases/v0/experiment.yaml (hunk 10 above)
Comment (wes-turner) [reply schemas/test_cases/v0/experiment.yaml]: It became obvious. I hadn't understood that the PR was a WIP you were going to change before committing.

File: e2e_tests/tests/config.py (hunk 4 above)
Comment (wes-turner) [reply e2e_tests/tests/config.py]: > If so, then could we amend this PR to have the final path?

I hadn't been sure about this part of the rollout plan. I see you've done exactly this. Thanks!

File: .circleci/config.yml (hunk 3 above)
Comment (wes-turner) [reply .circleci/config.yml]: Thanks!

File: requirements.txt (hunk 6 above)
Comment (wes-turner) [reply requirements.txt:24]: https://hpe-aiatscale.atlassian.net/browse/MLG-445

File: e2e_tests/tests/requirements.txt (hunk 1 above)
Comment (wes-turner) [reply e2e_tests/tests/requirements.txt]: New comment/pin look good.

**Review by wes-turner — APPROVED** [review]
